File: scope104/src/main/java/com/petro/scope104/data/db/UserDataSourceDbImpl.java
package com.petro.scope104.data.db;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;

import com.petro.scope104.data.UserDataSource;
import com.petro.scope104.data.db.dao.CountryDao;
import com.petro.scope104.data.db.dao.UserDao;
import com.petro.scope104.data.db.entity.CountryEntity;
import com.petro.scope104.data.db.entity.UserEntity;
import com.petro.scope104.domain.entity.WorkerEntity;
import com.petro.scope104.presentation.list.Gender;

import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Observable;

public class UserDataSourceDbImpl implements UserDataSource {
    private final UserDao userDao;
    private final CountryDao countryDao;
    private final DbConverter dbConverter;

    @Inject
    public UserDataSourceDbImpl(UserDao userDao, CountryDao countryDao, DbConverter dbConverter) {
        this.userDao = userDao;
        this.countryDao = countryDao;
        this.dbConverter = dbConverter;
    }

    @Override
    public Observable<List<WorkerEntity>> loadUsers(int pageNumber, int pageSize, @NonNull Gender gender, @NonNull List<String> countries) {
        Observable<List<UserEntity>> usersResult;
        Boolean isMale = null;
        if (gender == Gender.MALE) {
            isMale = true;
        } else if (gender == Gender.FEMALE) {
            isMale = false;
        }
        if (countries.isEmpty()) {
            usersResult = userDao.loadUsers(pageNumber, pageSize, isMale);
        } else {
            usersResult = userDao.loadUsers(pageNumber, pageSize, isMale, countries);
        }
        return usersResult.map(input -> input.stream().map(dbConverter::mapDatabaseToUi).collect(Collectors.toList()));
    }

    @Override
    public void saveUsers(List<WorkerEntity> users) {
        List<UserEntity> listDB = users.stream().map(dbConverter::mapUiToDatabase).collect(Collectors.toList());
        List<CountryEntity> countryList = users.stream().map(WorkerEntity::getNat).map(dbConverter::mapUiToCountryEntity).collect(Collectors.toList());
        userDao.insert(listDB);
        countryDao.insert(countryList);
    }
}

File: scope104/src/main/kotlin/com/petro/scope104/data/db/UserDataSourceDbImpl.kt
package com.petro.scope104.data.db

import com.petro.scope104.data.UserDataSource
import com.petro.scope104.data.db.dao.CountryDao
import com.petro.scope104.data.db.dao.UserDao
import com.petro.scope104.data.db.entity.UserEntity
import com.petro.scope104.domain.entity.WorkerEntity
import com.petro.scope104.presentation.list.Gender
import io.reactivex.rxjava3.core.Observable
import java.util.stream.Collectors
import javax.inject.Inject

class UserDataSourceDbImpl @Inject constructor(
    private val userDao: UserDao,
    private val countryDao: CountryDao,
    private val dbConverter: DbConverter
) : UserDataSource {
    override fun loadUsers(
        pageNumber: Int,
        pageSize: Int,
        gender: Gender,
        countries: List<String>
    ): Observable<List<WorkerEntity>> {
        val usersResult: Observable<List<UserEntity>>
        var isMale: Boolean? = null
        if (gender == Gender.MALE) {
            isMale = true
        } else if (gender == Gender.FEMALE) {
            isMale = false
        }
        if (countries.isEmpty()) {
            usersResult = userDao.loadUsers(pageNumber, pageSize, isMale)
        } else {
            usersResult = userDao.loadUsers(pageNumber, pageSize, isMale, countries)
        }
        return usersResult.map<List<WorkerEntity>> { input: List<UserEntity> ->
            input.stream()
                .map { userEntity: UserEntity -> dbConverter.mapDatabaseToUi(userEntity) }
                .collect(Collectors.toList())
        }
    }

    override fun saveUsers(users: List<WorkerEntity>) {
        val listDB = users.stream()
            .map { workerEntity: WorkerEntity -> dbConverter.mapUiToDatabase(workerEntity) }
            .collect(Collectors.toList())
        val countryList = users.stream().map { obj: WorkerEntity -> obj.nat }
            .map { countryCode: String? ->
                dbConverter.mapUiToCountryEntity(
                    countryCode!!
                )
            }.collect(Collectors.toList())
        userDao.insert(listDB)
        countryDao.insert(countryList)
    }
}